### ISSUE TITLE
Question about indexing logic for chosen/rejected pairs

### ISSUE THREAD

**Issue by Weiww-Xu** [YangRui2015/Generalizable-Reward-Model#9]
Hi~ Thank you for sharing the implementation of GRM! I'm trying to integrate the GRM loss into the reward model training module of [Llama-factory](https://github.com/hiyouga/LLaMA-Factory/tree/main), and I've encountered a question.

In the compute_loss method of your code, I noticed the following indexing logic for extracting chosen and rejected samples:
```
jidx = torch.arange(0, bsz, 2)  # chosen_ids
kidx = jidx + 1                 # rejected_ids
```
However, based on standard practices in contrastive learning (e.g., DPO), I think the indexing to pair each chosen sample with its corresponding rejected sample would be like:
```
jidx = torch.arange(batch_size)         # [0, 1, 2, ..., batch_size-1]
kidx = jidx + batch_size                # [batch_size, batch_size+1, ..., 2*batch_size-1]
```

For example, the compute_loss method in LLaMA-Factory explicitly splits the batch into two halves:
```
def compute_loss(
        self, model: "PreTrainedModel", inputs: dict[str, "torch.Tensor"], return_outputs: bool = False, **kwargs
    ) -> Union["torch.Tensor", tuple["torch.Tensor", list["torch.Tensor"]]]:
        r"""Compute pairwise loss. The first n examples are chosen and the last n examples are rejected.

        Subclass and override to inject custom behavior.

        Note that the first element will be removed from the output tuple.
        See: https://github.com/huggingface/transformers/blob/v4.40.0/src/transformers/trainer.py#L3842
        """
        _, _, values = model(**inputs, output_hidden_states=True, return_dict=True, use_cache=False)
        batch_size = inputs["input_ids"].size(0) // 2
        chosen_masks, rejected_masks = torch.split(inputs["attention_mask"], batch_size, dim=0)
        chosen_rewards, rejected_rewards = torch.split(values, batch_size, dim=0)
```

Could you explain why this indexing logic is used?

#
https://github.com/YangRui2015/Generalizable-Reward-Model/blob/3c31be2da0c6878300a01878cac6b623ba11de0d/reward_models/grm_reward_trainer.py#L109

**Comment by Weiww-Xu** [comment]
I got it after reviewing the dataset building functions!
